File: csharp/variable-length-quantity/VariableLengthQuantity.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class VariableLengthQuantity
{
    public static uint[] Encode(uint[] numbers) => numbers.SelectMany(n => AsVLQ(n)).ToArray();

    private static IEnumerable<uint> AsVLQ(uint source, bool isFirst = true)
    {
        if(source > 127)
            foreach(var next in AsVLQ(source >> 7, false))
                yield return next;

        var result = source & 127;
        yield return isFirst ? result : result + 128;
    }

    public static uint[] Decode(uint[] bytes) => FromVLQ(bytes).ToArray();

    private static IEnumerable<uint> FromVLQ(uint[] bytes)
    {
        if(bytes.Last() > 127)
            throw new InvalidOperationException();

        uint sum = 0;
        foreach(var val in bytes)
            if(val > 127)
            {
                sum = sum << 7;
                sum += val - 128;
            }
            else
            {
                sum = sum << 7;
                yield return sum + val;
                sum = 0;   
            }
    }
}

File: fsharp/variable-length-quantity/VariableLengthQuantity.fs
module VariableLengthQuantity

let rec asVLQ isFirst (number : uint32) =
    [
        if number > 127u then
            yield! asVLQ false (number >>> 7)
        
        let result = number &&& 127u

        if isFirst then yield byte result
        else yield byte <| result + 128u
    ]

let fromVLQ bytes =
    if List.last bytes > 127uy then None
    else
        bytes 
            |> List.fold (fun (results, current) byte ->
                let value = uint32 byte
                if value > 127u then
                    let newCurrent = (current <<< 7) + (value - 128u)
                    results, newCurrent
                else
                    let newResult = (current <<< 7) + value
                    (uint32 newResult)::results, 0u) ([], 0u)
            |> fst |> List.rev |> Some

let encode (numbers: uint32 list): byte list =
    numbers |> List.collect (asVLQ true)

let decode (bytes: byte list) =
    bytes |> fromVLQ